[ETK/error-reporting] Activate sentry for a segment of users

fullofcaffeine opened the pull request:
## Changes proposed in this Pull Request

Load Sentry in the error-reporting ETK module that applies to wp-admin + Gutenberg, but only activate it for a segment of users (1% for now, using a simple modulus calc), and a12s' blogs that have the `error-reporting-use-sentry` sticker* applied. Keep activating our homebrew error-reporter for all other users.

*_Add the sticker with `wp --url=<your-wpcom-test-site-url> blog-stickers add --sticker=error-reporting-use-sentry --who=<your-wpcom-username>`. To remove just swap the `add` command with `remove`._

The goal is to experiment with Sentry and test its feasibility, while leaving our homebrew error-reporting active for all others. We're still using the same ETK module to do so (as it can make it easier to ship it to WoA later).

If everything goes well, we might deprecate our homebrew solution in favor of Sentry. More details: p9oQ9f-Ay-p2#comment-1075 and p9oQ9f-zY-p2#comment-1055 under the `Next` section 🤞🏻 

## Testing instructions

⚠️ Before testing any of the following scenarios, be sure to activate jsconcat in your WPCOM sandbox (apply this diff: D63665-code). This will start serving files from CDN domains (**i.e `s0.wp.com` -- be sure to sandbox this, too!**) which gets us closer to a production environment and allows us to test for CORS issues (for more details, see this systems-request: pMz3w-cCq-p2) ⚠️ 

### Happy-path: Sentry works when activated, verify it correctly captures JS errors on the editor page

1. Apply the `error-reporting-user-sentry` sticker to your sandboxed test blog (as an a11n, you can apply the `error-reporting-use-sentry` sticker to a blog (see command above) to instruct this module to activate Sentry);
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
2. You can then trigger an error manually [like this](https://github.com/Automattic/wp-calypso/pull/49485#issuecomment-828029870) from the editor page. Be sure to select the wp-admin iframe context (Gutenframe) if you're running the site through Calypso, or the error will be reported to the Calypso error reporter, instead!
3. Login to the A8C Sentry account and verify that the error has been captured in the `wpcom-gutenberg-wp-admin` project. Follow the instructions here: PCYsg-ldt-p2if you're not sure how to create your account and login).
4. Test "head" errors. These are errors that might happen before the error reporter (in this case, Sentry) has been loaded. In this case, it should be caught by and stored in a window-wise array, and then once the error reporting module is loaded - and after it activates Sentry  - it will then iterate through all of them and send them over the wire. There's currently some test code that will generate some exceptions for you. Verify that Sentry has 3 recent `Head error ${count} from the top-level document` errors shortly after you loaded the page.

### Additional scenarios

---

**UPDATE**: I [added](https://github.com/Automattic/wp-calypso/pull/54257/commits/05664a223a2d6bb04d4e2bb9044344373cc35f62) unit tests for the modulus segmentation/activation logic. See [here](https://github.com/Automattic/wp-calypso/pull/54257/commits/05664a223a2d6bb04d4e2bb9044344373cc35f62). I encourage to test these manually, too, since these are not integration tests, but we already have some further confidence in the core activation logic.

To run the tests:

`apps/editing-toolkit (try/sentry-for-a-segment-of-users●)$ yarn test:php`

@noahtallen Gave a good primer in this thread (slack): p1620858708150500-slack-C7YPUHBB2.

---

#### When loading a blog without the `error-reporting-use-sentry` applied as an a11n, the homebrew error is activated
1. Make sure your blog doesn't have the aforementioned sticker, remove it if necessary;
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating homebrew error reporting!`, which means our version of error-reporting has been loaded instead _(`homebrew` here has nothing to do with the packaging tool, but is a way to refer to our own error-reporting solution, I'm trying to come up with a better term :))_

#### When loading a blog with the `error-reporting-use-sentry` applied as an a11n, Sentry is activated
1. Apply the `error-reporting-user-sentry` to your sandboxed test blog;
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating Sentry!`
4. You can refer to the happy-path scenario above to further test Sentry and verify that it's capturing the errors.

#### Sentry is not activated for non-a12s:
1. In your sandbox, modify the `is_automattician` function to always return `false` (search for `function is_automattician` in opengrok to find where it's defined)
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating homebrew error reporting!`, which means our version of error-reporting has been loaded instead.

#### Test the modulus segmentation logic works and that it loads the correct error reporter for each segment
1. Find out your user_id, you can output it using l( get_current_user_Id ) somewhere in the index.php for the error-reporting ETK module;
2. Let’s say your user_id is 7922883, this means you’d fall into the 84% segment. The default value in the test PR is `10`, so it should not load Sentry for you by default (it'd only load ids ending in 0-9). Try changing the `$current_segment` variable's value [here](https://github.com/Automattic/wp-calypso/blob/try/sentry-for-a-segment-of-users/apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php#L54) to `84`, and then you should see `Activating Sentry` in the console.
3. Then change it to < 84 and you should see `Activating homebrew error reporting!` instead.

## Todo

- [ ] ~~Move the Sentry account data over to the config, in order to have different environments (staging, prod) - for now we'd probably only activate it on prod, though. See discussion: p1626280547282100/1625249660.115200-slack-C7YPUHBB2.~~

## Outstanding questions
- [ ] ~~Should we mutually activate each error reporting solution or would it make sense to activate both Sentry and our our error reporting at the same time?~~
- [x] What Sentry account to use? We can use the one I created using my A8C email (search for `Test Sentry account for WPCOM`  in the secret store) and move it to a Sentry-specific mailbox. This would require a systemsrequest to ask for the mailbox to be created.

fullofcaffeine committed:
Activate sentry for a segment of users, and a12s if the blog has the `error-reporting-use-sentry` sticker

fullofcaffeine committed:
Add debug output

fullofcaffeine committed:
Improve how data is passed from php and handled in the client js

scinos commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js:
```diff
@@ -3,46 +3,80 @@
+import { Integrations } from '@sentry/tracing';
+
+const dataFromPHP = window.dataFromPHP;
+const shouldActivateSentry =
```
I think this can be simplified to `window.dataFromPHP?.shouldActivateSentry === 'true'`

fullofcaffeine committed:
Simplify expression

fullofcaffeine commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js:
```diff
@@ -3,46 +3,74 @@
+const shouldActivateSentry = window.dataFromPHP?.shouldActivateSentry === 'true';
+
+function activateSentry() {
+	console.debug( '[error-reporting] Activating Sentry!' );
```
This and the other `console.debug` will be removed before merging.

fullofcaffeine committed:
Add test for the error reporter activation logic

fullofcaffeine committed:
Fix test name and improve comment

fullofcaffeine committed:
Fix stubbing for `is_automattician`

The previous implementation created a chicken-and-egg situation with the test env and broke the logic in production.

In the ETK PHP test env, `is_automattician` is not available globally. At first, I defined it under the `A8C\FSE\ErrorReporting` namespace in the test. This worked for the stubbing purposes, but required me to change the `function_exists` call in the implementation to include the namespace in order for it to work as expected in the test env. I thought that in production, this check would work, too, but it didn't. The function is defined globally in production and the check for it under that namespace failed, meaning that the `should_activate_sentry` always returned `false` in production.

The solution was to remove that `function_exists` check which was creating more trouble than it's worth, and make sure the function was available in the test env, for all tests.

fullofcaffeine committed:
Change to the definitive Sentry DSN

fullofcaffeine committed:
Change initial user test segment to 1%

fullofcaffeine committed:
Dedup packages

Using `npx yarn-deduplicate`.

 # Changesto be committed:

fullofcaffeine committed:
Capture head errors for Sentry too

fullofcaffeine committed:
Add debug code for head errors, will be reverted before merge

fullofcaffeine commented:
I just realized that I forgot about "head errors" (errors that happen before the error reporter loads) for Sentry. I added support for it in this commit (now it's a common logic for both error reporters): https://github.com/Automattic/wp-calypso/pull/54257/commits/7548e3d9857cd7ffe23bd2e0328b16a6a73ccfd4. I also added some test code to make it easier to test head errors (will be removed before merging): https://github.com/Automattic/wp-calypso/pull/54257/commits/021aa99718da2a33a4c9270c790f2ace642abe3a. Will be updating the testing scenarios in the description, too.

fullofcaffeine committed:
Improve error head capture logic for Sentry

Sentry.captureException doesn't return a Promise so it doesn't converge to the same type in order to have a single `reportError` that is awaited by `Promise.allSettled`. Sentry requires `Sentry.flush()` to be called if you want to guarantee that all events are pushed. This in turns returns a Promise. Better to just have separate logic for each error reporter, in this case.

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php:
```diff
@@ -40,45 +48,94 @@ function add_crossorigin_to_script_els( $tag ) {
+
+	wp_localize_script(
+		$script_id,
+		'dataFromPHP',
```
I think we should make this more unique, since we're setting a variable in the global context -- theoretically, any plugin in PHP could decide to use this name for their data and it might conflict.

fullofcaffeine replied:
Ah, good point. I'll add a prefix 👍🏻

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php:
```diff
@@ -40,45 +48,94 @@ function add_crossorigin_to_script_els( $tag ) {
+ */
+function should_activate_sentry( $user_id, $blog_id ) {
+	return ( is_automattician( $user_id ) && has_blog_sticker( 'error-reporting-use-sentry', $blog_id ) )
+		|| ( ! is_automattician( $user_id ) && user_in_sentry_test_segment( $user_id ) );
```
Does it make sense to just have it be `has_blog_sticker || user_in_sentry_test_segment`?

fullofcaffeine replied:
🤔 I guess we could, but there would be a very small probability of an a11n falling into the `user_sentry_test_segment` if it doesn't have the sticker, and forcedly have Sentry activated even when semantically it should have the "classic/homebrew" error reporting (because it's an a11n without the sticker to use Sentry).

fullofcaffeine replied:
To add to that: I don't think simplifying the expression as you suggest would be bad and that the possible consequences are a big deal, but I also think it's fine to leave as is since this logic is ephemeral: it'll be removed as soon as we decide what path to go: Sentry or Classic (or both).

noahtallen replied:
Sounds good!

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js:
```diff
@@ -1,45 +1,75 @@
+		);
 	};
 
-	return (
```
Should we keep the current error reporting around for the short term? (Not sure if we decided to keep it yet or not?)

fullofcaffeine replied:
Yes, it'll be around, these removals are just moved lines. It will still be activated for the majority of customers.

noahtallen replied:
Ah, my bad! I didn't read closely enough and thought the code was being deleted, not moved.

fullofcaffeine committed:
Make the name of the obj used to pass data from PHP to JS more unique

It's in window and it was a bit too general, so another plugin could maybe add something that could match and replace it (or it could affect an existing reference).

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js:
```diff
@@ -2,7 +2,7 @@ import * as Sentry from '@sentry/browser';
 import apiFetch from '@wordpress/api-fetch';
 
-const shouldActivateSentry = window.dataFromPHP?.shouldActivateSentry === 'true';
+const shouldActivateSentry = window.A8C_FSE_ErrorReporting_Config?.shouldActivateSentry === 'true';
```
I would switch FSE for ETK, but that looks a lot better!

noahtallen reviewed (commented):
I can confirm that the head errors are getting sent to Sentry, but the custom errors unfortunately don't show up... Even when I set the frame context to post.php or post-new.php. 🤔 Any tips?

Update: we realized "custom errors" only worked in wp-admin, not gutenframe. We solved the problem by removing a legacy error handler which was overwriting the new error handler given by sentry (see #54912)

fullofcaffeine committed:
FSE -> ETK

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php:
```diff
@@ -40,45 +48,95 @@ function add_crossorigin_to_script_els( $tag ) {
  */
-function activate_error_reporting() {
+function setup_error_reporting() {
 	add_action( 'admin_print_scripts', __NAMESPACE__ . '\head_error_handler' );
```
Out of curiosity, would we want to set this to priority such that it is the first script executed by the browser? Otherwise, is there a chance that some scripts start getting executed (and throwing errors) before this is even available?

fullofcaffeine replied:
That's a good point! Thanks for bringing that up. I guess a scenario where another header script is printed/enqueued before this one is possible 🤔. I'm not super familiar with how the priority algorithm works, but the [documentation](https://developer.wordpress.org/reference/functions/add_action/) for `add_action`'s `$priority` parameter says:


>  (int) (Optional) Used to specify the order in which the functions associated with a particular action are executed. Lower numbers correspond with earlier execution, and functions with the same priority are executed in the order in which they were added to the action. 


So I'd say setting it to zero would do it?

fullofcaffeine replied:
I'll set it to `0` and do some smoke testing.

fullofcaffeine replied:
I saw no difference in the order the script element is added to the page when setting the priority to `0`, but I guess it's a good idea to leave it at it, given the default is `10`. It's actually already added as early as possible, though there is a snippet before it even if I set it to `0`:

![Screenshot from 2021-07-27 12-13-45](https://user-images.githubusercontent.com/81248/127198517-fc38aa82-3b6d-44ed-b7dc-d0c93fff2a9f.png)

noahtallen replied:
Nice, that's good to know!

fullofcaffeine committed:
Remove debug code

fullofcaffeine committed:
Spelling fix

noahtallen reviewed (approved):
This tests well for me!

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php:
```diff
@@ -40,45 +40,95 @@ function add_crossorigin_to_script_els( $tag ) {
+ * Used to check if the sticker is applied if user is A8C.
+ */
+function should_activate_sentry( $user_id, $blog_id ) {
+	return ( is_automattician( $user_id ) && has_blog_sticker( 'error-reporting-use-sentry', $blog_id ) )
```
Neither of these functions will be available on Atomic sites. We're probably safe right now, since the function isn't called on Atomic, but it might be safer to add another gate here

fullofcaffeine replied:
Yes, I'll review its use in the follow-up AT PR.

fullofcaffeine committed:
Set the priority of the `admin_print_scripts` action to 0 to make sure it is added as early as possible
